Bugfix: Correct first-run free space checks

luke-jr opened the pull request:
It's not clear what `m_assumed_*_size` are actually set based on, but historically it was in GB, not GiB, and that's still used in the GUI which is more user-facing.

Could just as easily change the GUI if GiB is preferred.

luke-jr committed:
Bugfix: init: For first-run disk space check, advise user of correct pruned size rather than full blockchain size

luke-jr committed:
Bugfix: init: Correct conversion of AssumedBlockchainSize to use GB

luke-jr committed:
doc/release-process: Correct m_assumed_*_size to GB

Sjors commented:
I'm not sure what the latest convention is, cc @hebasto. A few years ago #15163 made it so that RPC and most code uses KiB / MiB / GiB, but the GUI uses KB / MB / GB.

sydgren commented on src/init.cpp:
```diff
@@ -1622,7 +1622,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     // On first startup, warn on low block storage space
     if (!fReindex && !fReindexChainState && chain_active_height <= 1) {
-        uint64_t assumed_chain_bytes{chainparams.AssumedBlockchainSize() * 1024 * 1024 * 1024};
+        uint64_t assumed_chain_bytes{chainparams.AssumedBlockchainSize() * 1'000'000'000};
```
```suggestion
        uint64_t assumed_chain_bytes{chainparams.AssumedBlockchainSize() * pow(2, 30)};
```

```suggestion
        uint64_t assumed_chain_bytes{chainparams.AssumedBlockchainSize() * 1024 * 1024 * 1024};
```

Storage multipliers are defined as powers of 2 in increments of 10. Either use `pow(2, 30)` or the original `1024 * 1024 * 1024` for correct GiB conversion.

Any of the above suggestions would work. I don't understand the change in math.

fanquake commented:
@hebasto can you follow up given the gui / translation Qs here. This also needs a rebase.

hebasto commented:
Indeed, there is inconsistency in GB/GiB unit usage.

For instance, the value of the `m_assumed_blockchain_size` variable in GiB is used with the "GB" units in the user-faced messages here: https://github.com/bitcoin/bitcoin/blob/45f757c72672fc351bf669e731744f2f5e269233/src/init.cpp#L1707 and here: https://github.com/bitcoin/bitcoin/blob/45f757c72672fc351bf669e731744f2f5e269233/src/qt/forms/intro.ui#L206

Concept ACK.

---

It would be nice to mention the second commit changes in the PR description, no?

---

> Could just as easily change the GUI if GiB is preferred.

The GUI now uses "GB" (SI prefix) as a unit, based on the mindset of non-technical/non-CS users.

@GBKS What are modern guidelines in this regard?

---
@Sjors 
> I'm not sure what the latest convention is, cc @hebasto. A few years ago #15163 made it so RPC, code and GUI all use KiB / MiB / GiB.
> 
> (though for pruning it seems the config file uses MiB, but the GUI converts it to GB - yet there are translated error strings using `MiB`, confusing...)

For the GUI, the conversion to/from GB is still used: https://github.com/bitcoin/bitcoin/blob/45f757c72672fc351bf669e731744f2f5e269233/src/qt/optionsmodel.h#L26-L34

GBKS commented:
> The GUI now uses "GB" (SI prefix) as a unit, based on the mindset of non-technical/non-CS users.

`GB` is standard for general audiences, and `GiB` for technical, scientific, or engineering contexts. I'd recommend keeping `GB` in the GUI.

I am curious about use cases where it is essential for the user to know the `GiB` value in order to make decisions.

murchandamus reviewed (commented):
All of these changes seem consistent to me.

Concept ACK on generally using GB to refer to 1,000,000,000 bytes and GiB to refer to 1024³ bytes everywhere.

BrandonOdiwuor reviewed (commented):
Concept ACK using `GB` to represent 1000<sup>3</sup> bytes for consistency with GUI

sedited reviewed (approved):
ACK c452d6c1efe253daa91f1269d378ff82971315f0

luke-jr committed:
util: add overflow-safe `CeilDiv` helper

Introduce `CeilDiv()` for integral ceiling division without the typical `(dividend + divisor - 1) / divisor` overflow.
`CeilDiv()` asserts non-negative arguments and a non-zero divisor.
Add unit tests covering return type deduction, max-value behavior, and divisor checks.

sedited commented:
@luke-jr ping for rebase. Can also drop the first commit now.

sedited commented:
I don't think Luke is going to come around to this patch anytime soon, so I'm closing this again. @l0rinc do you want to take this over, since you've done some similar work recently?

l0rinc commented:
Revived in https://github.com/bitcoin/bitcoin/pull/35948 using the existing `CeilDiv` and added new `_GB` and `_MB` literals for consistency.
